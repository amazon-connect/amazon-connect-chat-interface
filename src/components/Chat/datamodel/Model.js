// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

export const PARTICIPANT_MESSAGE = "MESSAGE";
export const CHAT_EVENT = "EVENT";
export const ATTACHMENT_MESSAGE = "ATTACHMENT";

export const PARTICIPANT_TYPES = {
    AGENT: "AGENT",
    CUSTOMER: "CUSTOMER"
};

export const ContentType = {
    EVENT_CONTENT_TYPE: {
        TYPING: "application/vnd.amazonaws.connect.event.typing",
        PARTICIPANT_JOINED: "application/vnd.amazonaws.connect.event.participant.joined",
        PARTICIPANT_LEFT: "application/vnd.amazonaws.connect.event.participant.left",
        TRANSFER_SUCCEEDED: "application/vnd.amazonaws.connect.event.transfer.succeeded",
        TRANSFER_FAILED: "application/vnd.amazonaws.connect.event.transfer.failed",
        CONNECTION_ACKNOWLEDGED: "application/vnd.amazonaws.connect.event.connection.acknowledged",
        CHAT_ENDED: "application/vnd.amazonaws.connect.event.chat.ended"
    },
    MESSAGE_CONTENT_TYPE: {
        TEXT_PLAIN: "text/plain",
        APPLICATION_PDF: "application/pdf",
        IMAGE_JPG: "image/jpeg",
        IMAGE_PNG: "image/png",
        APPLICATION_DOC: "application/msword",
        APPLICATION_DOCX: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        APPLICATION_XLS: "application/vnd.ms-excel",
        APPLICATION_XLSX: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        TEXT_CSV: "text/csv",
        APPLICATION_PPT: "application/vnd.ms-powerpoint",
        APPLICATION_PPTX: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        AUDIO_WAV: "audio/wav",
        AUDIO_X_WAV: "audio/x-wav", //Firefox
        AUDIO_VND_WAVE: "audio/vnd.wave", //IE
        INTERACTIVE_MESSAGE: "application/vnd.amazonaws.connect.message.interactive"
    }
};

ContentType.ATTACHMENT_CONTENT_TYPE = {
    TXT: ContentType.MESSAGE_CONTENT_TYPE.TEXT_PLAIN,
    PDF: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PDF,
    JPG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_JPG,
    PNG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_PNG,
    DOC: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOC,
    DOCX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOCX,
    XLS: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLS,
    XLSX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLSX,
    CSV: ContentType.MESSAGE_CONTENT_TYPE.TEXT_CSV,
    PPT: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPT,
    PPTX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPTX,
    WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_WAV,
    X_WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_X_WAV,
    VND_WAVE: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_VND_WAVE
};

//OpenXML content types do not show up in custom files list unless file extension is explicitly provided
export const ATTACHMENT_ACCEPT_CONTENT_TYPES = [
    ...Object.values(ContentType.ATTACHMENT_CONTENT_TYPE),
    ".docx",
    ".xlsx",
    ".pptx"
];

export const InteractiveMessageType = {
    LIST_PICKER: "ListPicker",
    TIME_PICKER: "TimePicker",
    PANEL: "Panel"
};

export const Status = {
    Sending: "Sending",
    SendSuccess: "SendSuccess",
    SendFailed: "SendFailed",
    Read: "Read"
    // Delivered?
};

export const AttachmentStatus = {
    APPROVED: "APPROVED",
    REJECTED: "REJECTED"
};

export const Direction = {
    Outgoing: "Outgoing",
    Incoming: "Incoming"
};

export const AttachmentErrorType = {
    AccessDeniedException: "AccessDeniedException",
    InternalServerException: "InternalServerException",
    ThrottlingException: "ThrottlingException",
    ValidationException: "ValidationException",
    ServiceQuotaExceededException: "ServiceQuotaExceededException",
    ResourceNotFoundException: "ResourceNotFoundException",
    ConflictException: "ConflictException"
};
export class TransportDetails {
    constructor(input) {
        var args = input || {};
        this.direction = args.direction;
        this.readTime = args.readTime;
        this.status = args.status;
        this.sentTime = args.sentTime;
    }
}

export class ItemDetails {
    constructor(input) {
        var args = input || {};
        this.id = args.id;
        this.type = args.type;
        this.content = args.content;
        this.displayName = args.displayName;
        this.participantId = args.participantId;
        this.participantRole = args.participantRole;
        this.transportDetails = args.transportDetails;
    }
}
