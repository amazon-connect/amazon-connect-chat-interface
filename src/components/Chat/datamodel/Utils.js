// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import {
  Status,
  Direction,
  ItemDetails,
  TransportDetails,
  ContentType,
  PARTICIPANT_TYPES
} from "./Model";

function isRecognizedEvent(eventName) {
  var values = Object.values(ContentType.EVENT_CONTENT_TYPE);
  for (var index = 0; index < values.length; index++) {
    if (values[index] === eventName) {
      return true;
    }
  }
  return false;
}

function getContent(item) {
  var content = {};
  content.data = item.Content;
  content.type = item.ContentType;
  return content;
}

function createItemFromIncoming(item, thisParticipant) {
  var transcriptItem = new ItemDetails();
  var transportDetails = new TransportDetails();
  transcriptItem.id = item.Id;
  transcriptItem.type = item.Type;
  transcriptItem.content = getContent(item);
  transcriptItem.displayName = item.DisplayName;
  transcriptItem.participantId = item.ParticipantId;
  transcriptItem.participantRole = item.ParticipantRole;
  transportDetails.direction = 
      item.ParticipantRole === PARTICIPANT_TYPES.CUSTOMER
        ? Direction.Outgoing
        : Direction.Incoming;
  transportDetails.sentTime = 
    new Date(item.AbsoluteTime).getTime() / 1000;
  transportDetails.status = Status.SendSuccess;
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.version = 0;
  transcriptItem.Attachments = item.Attachments;
  return transcriptItem;
}

function createTranscriptItemFromSuccessResponse(oldTranscriptItem, response) {
  const newTranscriptItem = new ItemDetails();
  Object.assign(newTranscriptItem, oldTranscriptItem);

  if(response.data && response.data.Id){
    newTranscriptItem.id = response.data.Id;
  }

  newTranscriptItem.transportDetails = {
    ...oldTranscriptItem.transportDetails,
    status: Status.SendSuccess,
    sentTime: new Date(response.data.AbsoluteTime || Date.now()).getTime() / 1000
  };
  return newTranscriptItem;
}

function createOutgoingTranscriptItem(type, content, participant){
  const transcriptItem = new ItemDetails();
  const transportDetails = {};
  transcriptItem.type = type;
  transcriptItem.content = content;
  transcriptItem.participantId = participant.participantId;
  transcriptItem.participantRole = PARTICIPANT_TYPES.CUSTOMER;
  transcriptItem.displayName = participant.displayName;
  transportDetails.status = Status.Sending;
  transportDetails.direction = Direction.Outgoing;
  transportDetails.sentTime = _timestampNow();
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.id = _generateLocalId();
  transcriptItem.version = 0;
  return transcriptItem;
}

function createFailedItem(item, sentTime) {
  const clonedItem = new ItemDetails(item);
  clonedItem.transportDetails.status = Status.SendFailed;
  clonedItem.transportDetails.sentTime = sentTime;
  clonedItem.version = clonedItem.version + 1;
  return clonedItem;
}

function _generateLocalId() {
  var dt = new Date().getTime();
  var uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
    c
  ) {
    var r = (dt + Math.random() * 16) % 16 | 0;
    dt = Math.floor(dt / 16);
    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
  });
  return uuid;
}

function _timestampNow() {
  var date = new Date();
  var timestamp = date.getTime();
  return timestamp / 1000;
}

function createTypingParticipant(typingDataItem, thisParticipantId) {
  var participantTypingDetails = {};
  participantTypingDetails.participantId = typingDataItem.ParticipantId;
  participantTypingDetails.displayName = typingDataItem.DisplayName;
  var direction =
    thisParticipantId === typingDataItem.ParticipantId
      ? Direction.Outgoing
      : Direction.Incoming;
  participantTypingDetails.direction = direction;
  return participantTypingDetails;
}

function isAttachmentContentType(contentType) {
  return contentType && Object.values(ContentType.ATTACHMENT_CONTENT_TYPE).includes(contentType.toLowerCase());
}

var modelUtils = {
  createItemFromIncoming: createItemFromIncoming,
  createOutgoingTranscriptItem: createOutgoingTranscriptItem,
  createFailedItem: createFailedItem,
  createTypingParticipant: createTypingParticipant,
  isRecognizedEvent: isRecognizedEvent,
  createTranscriptItemFromSuccessResponse: createTranscriptItemFromSuccessResponse,
  isAttachmentContentType: isAttachmentContentType,
};

export { modelUtils };