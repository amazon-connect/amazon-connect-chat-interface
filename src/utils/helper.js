import { ContentType, InteractiveMessageType } from "../components/Chat/datamodel/Model";
import { INTERACTIVE_MESSAGE } from "../components/Chat/constants";
import * as DOMPurify from 'dompurify';

export function shouldDisplayMessageForType(correntType) {
  let isValid = false;
  for (let key in ContentType.MESSAGE_CONTENT_TYPE) {
    if (ContentType.MESSAGE_CONTENT_TYPE[key] === correntType) {
      isValid = true;
    }
  }
  return isValid;
}

export function getTimeFromTimeStamp(timeStamp) {
  return new Date(timeStamp).getTime();
}

export function createInteractiveMessagePayload(
  selectedElement,
  preIndex,
  nextIndex,
  listId,
  templateType,
  referenceId,
  isCarouselElem = false,
  listTitle,
  carouselTemplateId
) {
  if (isCarouselElem) {
    return {
        text: JSON.stringify({
          templateIdentifier: carouselTemplateId,
          listTitle,
          selectionText: selectedElement.title
        })
    }
  }

  const payload = { text: selectedElement.title };
  if (
    selectedElement.actionDetail &&
    (selectedElement.actionDetail === INTERACTIVE_MESSAGE.ACTIONS.SHOW_MORE ||
      selectedElement.actionDetail ===
        INTERACTIVE_MESSAGE.ACTIONS.PREVIOUS_OPTIONS)
  ) {
    const requestBody = {
      version: INTERACTIVE_MESSAGE.VERSION,

      data: {
        actionName: selectedElement.actionDetail,
        preIndex,
        nextIndex,
        listId,
        templateType,
        referenceId,
      },
      action: selectedElement.actionDetail,
    };
    const jsonStr = JSON.stringify(requestBody);
    payload.text = jsonStr;
    payload.type = ContentType.MESSAGE_CONTENT_TYPE.INTERACTIVE_RESPONSE;
  }
  return payload;
}

/**
 * Frontend validations for interactive messages
 *  - Client will always receive a valid template to render
 *  - Frontend will truncate fields
 *  - Upper limits not enforced by backend, beyond a 20k char limit for the entire message
 *
 * Documentation: https://docs.aws.amazon.com/connect/latest/adminguide/interactive-messages.html
 */
export const INTERACTIVE_MESSAGE_CONSTRAINTS = {
  [InteractiveMessageType.PANEL]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
    elementTitleCharLimit: 400
  },
  [InteractiveMessageType.LIST_PICKER]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
    elementTitleCharLimit: 400,
    elementSubtitleCharLimit: 400,
  },
  [InteractiveMessageType.TIME_PICKER]: {
    titleCharLimit: 400,
    subtitleCharLimit: 400,
  },
  [InteractiveMessageType.CAROUSEL]: {
    titleCharLimit: 400,
  },
  [InteractiveMessageType.QUICK_REPLY]: {
    titleCharLimit: 400,
    replyOptionCharLimit: 200,
  }
}

/**
 * Truncates a string for interactive message based on hard-coded constraints
 *
 * @param {string} str - input string to truncate (e.g. title, subtitle).
 * @param {string} InteractiveMessageType - interactive message templateType.
 * @param {string} fieldLimitKey - field key for contraint (e.g. titleCharLimit).
 * @returns {string} the truncated string.
 */
export const truncateStrFromCharLimit = (str, InteractiveMessageType, fieldLimitKey) => {
  const templateContraints = INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType] || {};
  const MAX_LENGTH = templateContraints[fieldLimitKey] || 0;

  if (!(str && typeof str === "string" && MAX_LENGTH)) {
    return "";
  }

  /**
   * Mitigate Interactive Message fields XSS vulnerabilites with `dompurify`
   *
   * React will always render these passed template field values as strings not HTML.
   * React auto-escapes by default unless using dangerouslySetInnerHTML.
   */
  const sanitizedStr = DOMPurify.sanitize(str);

  if (sanitizedStr.length <= MAX_LENGTH) {
    return sanitizedStr;
  } else {
    console.warn(`[${InteractiveMessageType} template] ${fieldLimitKey} of ${MAX_LENGTH} was exceeded`)
    return sanitizedStr.substring(0, MAX_LENGTH) + "...";
  }
}

/** 
 * Generates the url to fetch guides renderer
 */
export const constructGuidesRendererUrl = (instanceAlias, rendererVersion) => {
  if (!instanceAlias || !rendererVersion) {
    console.warn("[GuidesInChat] Unable to generate guides renderer url. Chat will not be able to render views");
    return '';
  }
  const url = `https://${instanceAlias}.my.connect.aws/connectwidget/static/views/renderer/${rendererVersion}/index.js`;
  return url;
}

/**
 * Insert script in head to fetch guides renderer if required
 * @param {object} props props that were passed with the init call for this widget
 */
export const setupGuidesRenderer = (props) => {
  const logger = connect.LogManager ? connect.LogManager.getLogger({ prefix: "ChatInterface-Chat" }): console;
  if (props.guidesInChat) {
    const version = props.guidesInChat.version || 'latest';
    const instanceAlias = props.guidesInChat.instanceAlias;
    if (instanceAlias) {
      const guidesRendererUrl = constructGuidesRendererUrl(instanceAlias, version);
      logger && logger.debug('[GuidesInChat] Using guides renderer url ',guidesRendererUrl);

      const script = document.createElement("script");
      script.src = guidesRendererUrl;
      document.head.appendChild(script);
    } else {
      logger && logger.warn('[GuidesInChat] Could not find necessary configuration to fetch renderer. Guides in chat may not render');
    }
  } else {
    logger &&logger.warn('[GuidesInChat] Configuration was not provided. Guides in chat may not render if used outside of connect communication widget');
  }
}
