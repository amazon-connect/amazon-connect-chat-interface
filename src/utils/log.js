/* Add your own logger function here
  var customizedLogger = {
    debug: (data) => {// customize logger function here},
    info: (data) => {// customize logger function here},
    error: (data) => {// customize logger function here}
  }
  "connect" is a global object that exported from src/components/Chat/amazon-connect-chat.js
  There are four levels available from connect.LogLevel - DEBUG, INFO, ERROR.
*/
export const config = {
    loggerConfig: {
        // Add your logger here
        // customizedLogger: customizedLogger,
        level: connect.LogLevel.INFO,
        useDefaultLogger: true
    },
};
