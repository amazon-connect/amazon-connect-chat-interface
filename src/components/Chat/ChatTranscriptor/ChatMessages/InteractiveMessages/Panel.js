// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React, { useState } from "react";
import { Button } from "connect-core";
import { Text } from "connect-core";
import {
  ReactiveImage,
  TextSection,
  ResponsesSection,
  Title,
  Subtitle,
} from "../InteractiveMessage";
import styled from "styled-components";
import PT from "prop-types";
import Linkify from "react-linkify";
import { createInteractiveMessagePayload } from "../../../../../utils/helper";

const ElementTitle = styled(Text)`
  font-weight: bold;
`;

const ImageContainer = styled.div`
  max-height: calc(95vw * (9 / 16));
  overflow: hidden;
  display: ${(props) => (props.showImage ? "flex" : "none")};
`;

const PanelButton = styled(Button)`
  display: flex;
  justify-content: center;
  width: 100%;
  max-width: none;
  border: ${({ theme }) => theme.globals.baseBorder};
  white-space: pre-line;

  &:hover {
    background: ${({ theme }) => theme.color.primary};
  }

  &:last-of-type {
    border-bottom-left-radius: ${({ theme }) => theme.spacing.mini};
    border-bottom-right-radius: ${({ theme }) => theme.spacing.mini};
    margin-bottom: 0;
  }
`;

Panel.propTypes = {
  content: PT.object.isRequired,
  addMessage: PT.func.isRequired,
};

export default function Panel({ content, addMessage, templateType }) {
  // assumptions: version 1, image data is URL. Guarenteed title exists, at least 1 element.
  const {
    title,
    subtitle,
    elements,
    imageData,
    imageDescription,
    preIndex,
    nextIndex,
    listId,
    referenceId,
  } = content;
  const [imageLoaded, setImageLoaded] = useState(false);

  function onImageLoad() {
    setImageLoaded(true);
  }

  function onItemClick(index) {
    const selectedElement = elements[index];
    const payload = createInteractiveMessagePayload(
      selectedElement,
      preIndex,
      nextIndex,
      listId,
      templateType,
      referenceId
    );
    addMessage(payload);
  }

  return (
    <>
      {imageData && (
        <ImageContainer showImage={imageLoaded}>
          <ReactiveImage
            imageSrc={imageData}
            imageDescription={imageDescription}
            onImageLoad={onImageLoad}
          />
        </ImageContainer>
      )}
      <TextSection>
        <Title>
          <Linkify properties={{ target: "_blank" }}>{title}</Linkify>
        </Title>
        {subtitle && (
          <Subtitle>
            <Linkify properties={{ target: "_blank" }}>{subtitle}</Linkify>
          </Subtitle>
        )}
      </TextSection>
      <ResponsesSection>
        <div>
          {elements.map((element, index) => (
            <PanelButton
              value={element.title}
              key={"element-" + index}
              onClick={() => {
                onItemClick(index);
              }}
            >
              <ElementTitle>{element.title}</ElementTitle>
            </PanelButton>
          ))}
        </div>
      </ResponsesSection>
    </>
  );
}
