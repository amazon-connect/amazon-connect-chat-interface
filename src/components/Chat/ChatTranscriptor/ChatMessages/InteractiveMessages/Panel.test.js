// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from 'react';
import { screen, render, fireEvent } from "@testing-library/react";
import { InteractiveMessageType } from "../../../datamodel/Model";
import { ThemeProvider } from "../../../../../theme";
import Panel from "./Panel";
import * as helpers from '../../../../../utils/helper';

const PANEL_CONSTRAINTS = helpers.INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType.PANEL];

const mockPanelContent = {
    title: "PanelTitle",
    subtitle: "PanelSubtitle",
    elements: [
        {
            title: "PanelElementTitle",
        },
        {
            title: "AnotherPanelElementTitle",
        }
    ],
    imageData: "PanelImageData",
    imageDescription: "PanelImageDescription"
};

describe("<Panel />", () => {
  let mockProps;

  function renderElement(props) {
    return render(
      <ThemeProvider>
        <Panel {...props} />
      </ThemeProvider>,
    );
  }

  beforeEach(() => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = { content: mockPanelContent, addMessage: addMessage, templateType: InteractiveMessageType.PANEL };
  });

  it("Style should match the snapshot", () => {
    const mockPanel = renderElement(mockProps);
    expect(mockPanel).toMatchSnapshot();
  });

  it("Should be able to use Panel", () => {
    renderElement(mockProps);

    expect(screen.getByText("PanelTitle")).toBeDefined()
    expect(screen.getByText("PanelSubtitle")).toBeDefined();
    expect(screen.getByAltText("PanelImageDescription")).toBeDefined();
    expect(screen.getByText("PanelElementTitle")).toBeDefined();
    expect(screen.getByText("AnotherPanelElementTitle")).toBeDefined();

    fireEvent.click(screen.getByText("PanelElementTitle"));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({
      text: "PanelElementTitle",
    });
  });

  it("Should create interactive selection message payload", () => {
    renderElement(mockProps);

    const createMsgPayloadSpy = jest.spyOn(helpers, 'createInteractiveMessagePayload');
    fireEvent.click(screen.getByText("PanelElementTitle"));
    expect(createMsgPayloadSpy).toHaveBeenCalledWith({"title": "PanelElementTitle"}, undefined, undefined, undefined, InteractiveMessageType.PANEL, undefined, undefined, "PanelTitle", undefined);

    createMsgPayloadSpy.mockRestore();
  });

  it.each([
      ["title", PANEL_CONSTRAINTS.titleCharLimit],
      ["subtitle", PANEL_CONSTRAINTS.subtitleCharLimit]
  ])("should truncate a %s over the character limit", (fieldKey, charLimit) => {
      const longTitle = "LongTitle".repeat(150);
      const truncatedTitle = `${longTitle.substring(0, charLimit)}...`;

      const panelPickerLongTitle = {
          ...mockPanelContent,
          [fieldKey]: longTitle,
      };
      renderElement({ ...mockProps, content: panelPickerLongTitle });
      expect(() => screen.getByText(longTitle)).toThrow(
          "Unable to find an element"
      );
      expect(() => screen.getByText(truncatedTitle)).not.toThrow();
  });

  it.each([
      ["title", PANEL_CONSTRAINTS.elementTitleCharLimit],
  ])("should truncate an element %s over the character limit", (fieldKey, charLimit) => {
      const longElementTitle = "LongElementTitle".repeat(100);
      const truncatedElementTitle = `${longElementTitle.substring(0, charLimit)}...`;

      const panelPickerLongTitle = {
          ...mockPanelContent,
          elements: [...mockPanelContent.elements, {
            ...mockPanelContent.elements[0],
            [fieldKey]: longElementTitle
          }]
      };
      renderElement({ ...mockProps, content: panelPickerLongTitle });
      expect(() => screen.getByText(longElementTitle)).toThrow(
          "Unable to find an element"
      );
      expect(() => screen.getByText(truncatedElementTitle)).not.toThrow();

      // Should still send entire title to backend, instead of returning 'my selectio...'
      fireEvent.click(screen.getByText(truncatedElementTitle));
      expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
      expect(mockProps.addMessage).toHaveBeenCalledWith({
        text: longElementTitle,
      });
  });
});

describe("Panel XSS Mitigation", () => {
  let mockProps;
  const mockPanelXSSContent = {
    title: 'Title with script <script>alert("XSS attack!");</script>',
    subtitle: '<img src="x" onerror="alert(\'XSS attack!\');">',
    elements: [
      {
        title: '<a href="javascript:alert(\'XSS attack!\')">Click me</a>',
      },
      {
        title: '<input type="text" value="XSS attack!" onfocus="alert(\'XSS attack!\');">',
      },
      {
        title: '<div data-value="<img src=x onerror=alert(\'XSS attack!\')>"></div>',
      },
      {
        title: '<div style="background-image: url(\'javascript:alert(\'XSS attack!\');\')"></div>',
      }
    ],
  };
  beforeEach(() => {
    mockProps = {
      content: mockPanelXSSContent,
      addMessage: jest.fn(),
      templateType: InteractiveMessageType.PANEL,
    };
  });
  function renderElement(props) {
    return render(
      <ThemeProvider>
        <Panel {...props} />
      </ThemeProvider>
    );
  }

  it("should use DOMPurify to mitigate malicious XSS input", () => {
    renderElement(mockProps);
    expect(screen.getByText("Title with script")).toBeDefined();
    expect(screen.getByText("<img src=\"x\">", { exact: false })).toBeDefined();
    expect(screen.getByText("<a>Click me</a>", { exact: false })).toBeDefined();
    expect(screen.getByText("<input value=\"XSS attack!\" type=\"text\">", { exact: false })).toBeDefined();
    expect(screen.getByText("<div data-value=\"<img src=x onerror=alert('XSS attack!')>\"></div>", { exact: false })).toBeDefined();
    expect(screen.getByText("<div style=\"background-image: url('javascript:alert('XSS attack!');')\"></div>", { exact: false })).toBeDefined();
  });
});