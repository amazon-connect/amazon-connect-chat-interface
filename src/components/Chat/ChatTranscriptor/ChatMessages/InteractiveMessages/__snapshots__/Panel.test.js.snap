// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Style should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c3 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  color: inherit;
  margin: 0;
}

.c3:last-child {
  margin: 0;
}

.c7 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c7 > img,
.c7 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c4 {
  font-weight: bold !important;
}

.c5 {
  color: #555555;
}

.c1 {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c2 {
  padding: 14px;
  text-align: left;
}

.c6 {
  position: relative;
  white-space: pre-line;
  border-radius: 6px;
}

.c9 {
  font-weight: bold;
}

.c0 {
  max-height: calc(95vw * (9 / 16));
  overflow: hidden;
  display: none;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  max-width: none;
  border: 1px solid #DDDDDD;
  white-space: pre-line;
}

.c8:hover {
  background: #06adcb;
}

.c8:last-of-type {
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: 0;
}

<body>
    <div>
      <div
        class="c0"
      >
        <img
          alt="PanelImageDescription"
          class="c1"
          src="PanelImageData"
        />
      </div>
      <div
        class="c2"
      >
        <p
          class="c3 c4"
        >
          <span
            class="Linkify"
          >
            PanelTitle 
            <a
              href="https://www.amazon.com/"
              target="_blank"
            >
              https://www.amazon.com/
            </a>
          </span>
        </p>
        <p
          class="c3 c5"
        >
          <span
            class="Linkify"
          >
            PanelSubTitle 
            <a
              href="https://www.amazon.com/"
              target="_blank"
            >
              https://www.amazon.com/
            </a>
          </span>
        </p>
      </div>
      <div
        class="c6"
      >
        <div>
          <button
            class="c7 c8"
            data-selected="false"
            type="default"
            value="PanelElementTitle"
          >
            <p
              class="c3 c9"
            >
              PanelElementTitle
            </p>
          </button>
          <button
            class="c7 c8"
            data-selected="false"
            type="default"
            value="AnotherPanelElementTitle"
          >
            <p
              class="c3 c9"
            >
              AnotherPanelElementTitle
            </p>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iRbamj dZAEsV"
    >
      <img
        alt="PanelImageDescription"
        class="sc-cMljjf dqMdBl"
        src="PanelImageData"
      />
    </div>
    <div
      class="sc-jAaTju ldErGV"
    >
      <p
        class="sc-bZQynM fraGhu sc-jWBwVP NljNK"
      >
        <span
          class="Linkify"
        >
          PanelTitle 
          <a
            href="https://www.amazon.com/"
            target="_blank"
          >
            https://www.amazon.com/
          </a>
        </span>
      </p>
      <p
        class="sc-bZQynM fraGhu sc-brqgnP dLoklq"
      >
        <span
          class="Linkify"
        >
          PanelSubTitle 
          <a
            href="https://www.amazon.com/"
            target="_blank"
          >
            https://www.amazon.com/
          </a>
        </span>
      </p>
    </div>
    <div
      class="sc-jDwBTQ cICqDD"
    >
      <div>
        <button
          class="sc-gzVnrw fgSGc sc-jlyJG bLgwDh"
          data-selected="false"
          type="default"
          value="PanelElementTitle"
        >
          <p
            class="sc-bZQynM fraGhu sc-gPEVay fXmaZh"
          >
            PanelElementTitle
          </p>
        </button>
        <button
          class="sc-gzVnrw fgSGc sc-jlyJG bLgwDh"
          data-selected="false"
          type="default"
          value="AnotherPanelElementTitle"
        >
          <p
            class="sc-bZQynM fraGhu sc-gPEVay fXmaZh"
          >
            AnotherPanelElementTitle
          </p>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
